### PR TITLE
fix: deserialize MessageType's other than Unknown

### PR DESCRIPTION
**Pull request by mkenigs**
456cc7b07052e84571d514b395609f608a962850 made MessageType an untagged enum so that the Unknown variant would deserialize correctly, but this meant that all strings would deserialize as Unknown rather than being deserialized into the correct variant. "constraints_too_tight" for example was being deserialized as `Unknown("constraints_too_tight")`.

The Unknown variant should be untagged, but the enum as a whole should not be.

### CODE REVIEW

File: cli/catalog-api-v1/src/error.rs [hunk 1]
```diff
@@ -14,5 +13,28 @@ pub enum MessageType {
...
     ConstraintsTooTight,
 
+    #[serde(untagged)]
     Unknown(String),
 }
+
+#[cfg(test)]
+mod tests {
+    #[test]
+    #[ignore = "useful when developing"]
```
Comment (mkenigs) [inline cli/catalog-api-v1/src/error.rs:23]: I can drop or run these if we prefer, but they'll be covered by re-generating mock data

File: cli/catalog-api-v1/src/error.rs [hunk 2]
```diff
@@ -14,5 +13,28 @@ pub enum MessageType {
     #[serde(rename = "constraints_too_tight")]
     ConstraintsTooTight,
 
+    #[serde(untagged)]
```
Comment (ysndr) [inline cli/catalog-api-v1/src/error.rs:16]: Its a bit of an unusual usage of `untagged` to me at least.
Why did you change this to the variant?

File: cli/catalog-api-v1/src/error.rs [hunk 3]
```diff
@@ -3,7 +3,6 @@ use std::hash::Hash;
 use serde::{Deserialize, Serialize};
```
Comment (ysndr) [inline cli/catalog-api-v1/src/error.rs:4]: suggestion: may be a good time to add some additional docs for this here:

```suggestion
/// Type of the error returned by the catalog API
/// Since we were unable to represent earlier error structures returned by the API
/// using the progenitor client generator,
/// errors are now serialized as a blob of values
/// (`context` in [crate::ResolutionMessageGeneral]).
///
/// The context may be parsed into a higher level structure later,
/// or ignored in which case the `message` field in [crate::ResolutionMessageGeneral]
/// is expected to provide a relevant fallback message.
```

File: cli/catalog-api-v1/src/error.rs (hunk 2 above)
Comment (mkenigs) [reply cli/catalog-api-v1/src/error.rs:16]: I don't exactly understand the question, but I wanted `"constraints_too_tight"` to deserialize to `ConstraintsTooTight` not `Unknown("constraints_too_tight")`

I found this usage via https://github.com/serde-rs/serde/issues/912#issuecomment-1868785603

Comment (ysndr) [reply cli/catalog-api-v1/src/error.rs:24]: oh what, it did deserialize as Unknown even though there's a variant for it 🤔
that's surprising, i guess that alone is a reason to keep the tests lol

nevermind then, untagged on the field seems to be the right use then

File: cli/catalog-api-v1/src/error.rs (hunk 1 above)
Comment (ysndr) [reply cli/catalog-api-v1/src/error.rs:31]: It think it would be useful to use this as a value explicitly, to signal its used as a value/constant.


```
let map: HashMap<String, MessageType> = serde_json::from_value(&json!({
 "known_type": "constraints_too_tight",
 "unknwon_type": "something unknown"
})).unwrap();

assert_eq!(map["known_type", MessageType::ConstraintsTooTight);
assert_eq!(map["unknown_type", MessageType::Unknown("something unknown".to_string()));

```

**Review by ysndr — APPROVED** [review]

File: cli/catalog-api-v1/src/error.rs (hunk 1 above)
Comment (mkenigs) [reply cli/catalog-api-v1/src/error.rs]: Merged the two tests into what you suggested
